packages/synapse-react-client/src/utils/hooks/useUploadFileEntity/useUploadFileEntities.ts

```diff
@@ -0,0 +1,434 @@
+          ),
+          progressCallback: progress => {
+            if (progress) {
+              setProgress(preparedFile.file, progress)
+            }
+          },
+          abortController: trackedProgress.abortController,
+          onMd5Computed: () => {
+            setIsUploading(preparedFile.file)
+          },
+        }
+
+        let newFileHandleId: string
+        if (
+          uploadDestination &&
+          instanceOfExternalObjectStoreUploadDestination(uploadDestination) &&
+          uploadDestination.endpointUrl != null
+        ) {
+          const fileHandle = await uploadFileDirectlyToS3({
+            ...uploadArgs,
+            bucketName: uploadDestination.bucket!,
+            endpoint: uploadDestination.endpointUrl,
+            keyPrefixUUID: uploadDestination.keyPrefixUUID!,
+            accessKey,
+            secretKey,
+          })
+          newFileHandleId = fileHandle.id!
+        } else {
+          const fileUploadComplete = await uploadFile(uploadArgs)
+
+          newFileHandleId = fileUploadComplete.fileHandleId
+        }
+
+        if (preparedFile.existingEntityId) {
+          // Update the existing entity
+          const entity =
+            await synapseClient.entityServicesClient.getRepoV1EntityId({
+              id: preparedFile.existingEntityId,
+            })
+          await updateEntityWithNewFile({
+            ...entity,
+            dataFileHandleId: newFileHandleId,
+          } as FileEntity)
+          // Mark file as done!
+          setComplete(preparedFile.file)
+        } else {
+          // else, it's a new file entity
+          const newFileEntity: FileEntity = {
+            parentId: preparedFile.parentId,
+            name: preparedFile.file.name,
+            concreteType: 'org.sagebionetworks.repo.model.FileEntity',
+            dataFileHandleId: newFileHandleId,
+          }
+          await createEntityWithNewFile(newFileEntity)
+          // Mark file as done!
+          setComplete(preparedFile.file)
+        }
+      } catch (e) {
+        console.error('File upload failed: ', e)
+        setFailed(preparedFile.file, e.message)
+      }
+    },
+    [
+      storageLocationId,
+      uploadDestination,
+      setProgress,
+      setIsUploading,
+      uploadFileDirectlyToS3,
+      accessKey,
+      secretKey,
+      uploadFile,
+      synapseClient.entityServicesClient,
+      updateEntityWithNewFile,
+      setComplete,
+      createEntityWithNewFile,
+      setFailed,
+    ],
+  )
+
+  /**
+   * Upload the list of prepared files to Synapse.
+   */
+  const startUpload = useCallback(
+    (...preparedFiles: FilePreparedForUpload[]) => {
+      clearPendingFiles()
+
+      const newTrackedUploadProgress = trackNewFiles(...preparedFiles)
+
+      // TODO: We already got the storage location from the root, but what if one of the sub-folders already existed
+      //   with a different storage location? should we use that storage location?
```

Comment:
Good question.  I think it's understandable if the web client does not have this sophisticated logic at this time (to recognize new folder storage locations during the upload of subfolder files)